rfd/0171-database-session-playback.md

```diff
@@ -0,0 +1,376 @@
+include the command result event:
+
+```proto
+// DatabaseFieldType enum for database data fields data types.
+enum DatabaseDataFieldType {
+  UNSPECIFIED = 0;
+  STRING = 1;
+  NUMBER = 2;
+  DATE = 3;
+  BINARY = 4;
+}
+
+// DatabaseDataFieldMetadata contains metadata for database data fields.
+message DatabaseDataFieldMetadata {
+  // Name is the field name.
+  string Name = 1;
+  // Type is the field type. Since the values are stored inside the event as
+  // strings, this type can convert the row values back into their original
+  // type. This information can be used to format the row values.
+  DatabaseDataFieldType Type = 2;
+}
+
+// DatabaseDataRow represents a single data row, containing a list field values.
+message DatabaseDataRow {
+  // Values the row field values.
+  // Value is the row field values. The values aren not expected to be the
+  // direct database type but their string representation, making it possible to
+  // display them without parsing. For example, binary data can be encoded into
+  // base64.
+  repeated string Values = 1;
+}
+
+// DatabaseResultData database results data, containing fields metadata and also
+// the data itself.
+message DatabaseResultData {
+  // FieldMetadata list of fields metadata.
+  repeated DatabaseDataFieldMetadata Fields = 1;
+  // Rows list of the field values.
+  repeated DatabaseDataRow Rows = 2;
+}
+
+// DatabaseSessionCommandResult represents the result of a user command. It is
+// expected that for each user command/query there will be a corresponding
+// result.
+message DatabaseSessionCommandResult {
+  // Metadata is a common event metadata.
+  Metadata Metadata = 1;
+  // User is a common user event metadata.
+  UserMetadata User = 2;
+  // SessionMetadata is a common event session metadata.
+  SessionMetadata Session = 3;
+  // Database contains database related metadata.
+  DatabaseMetadata Database = 4;
+
+  oneof Result {
+    // Status of the query execution. It is used when the command doesn't return
+    // any data.
+    Status status = 5;
+    // Data contains the fields and data of the query execution.
+    DatabaseResultData Data = 6;
+  }
+}
+```
+
+#### Examples
+
+* Audit event generated for a query that returns data (`SELET id, name FROM events;`):
+
+```json
+{
+  ...
+  "data": {
+    "fields": [
+      {"name": "id", "type": "INTEGER"},
+      {"name": "name", "type": "STRING"}
+    ],
+    "rows": [
+      ["1", "session.start"],
+      ["2", "session.end"]
+    ]
+  }
+}
+```
+
+* Query that only returns a success message (`UPDATE events SET ...`):
+
+```json
+{
+  ...
+  "status": {
+    "success": true,
+    "message": "3 rows affected"
+  }
+}
+```
+
+* Query that returns an error (`SELECT err;`):
+
+```json
+{
+  ...
+  "status": {
+    "success": false,
+    "error": "ERROR: column \"err\" does not exist"
+  }
+}
+```
+
+### Recording options
+
+Database session recording options will be defined per role. This configuration
```

Comment:
No global setting for the db recording mode? 

I don't see how a customer will benefit from a setup where `on` for one role but `full` for another role. Maybe `off` is good for Machine ID cases. But in general, I think global settings are more useful.

Reply:
Yeah, I considered making this global, but I couldn't decide where to place it so that we could have consistency across different databases:

- Placing at the db_service introduced a hard-to-keep configuration. If you have many DB nodes, it might get out of hand. Also, the session recordings could be inconsistent when different nodes can handle a database connection.
- Making it a "cluster-level" configuration (similar to `auth_serivice.session_recording`) seems to reduce the ability of administrators to decide which database would have the recording. This appears to be helpful, for example, if customers have dev/staging databases without the need to keep the recordings.